### PR TITLE
Added support for JWT generated by IDPs like Keycloak

### PR DESCRIPTION
**Pull request by mcharest-mcn**
### RSA Support
I was unable to use RSA256 to validate signed requests because the only supported encryption method was HMAC which is a symmetrical and I'm unable to exchange private key securely. In theory, it should also support the other asymmetrical key types, but I have not tested them.

To support RSA (and IDPs like Keycloak), I've added two new configuration parameters : PREST_JWT_WELLKNOWN and PREST_JWT_JWKS.

### PREST_JWT_WELLKNOWN
Takes an URL to the well-known configuration of your IDP (ex: https://accounts.google.com/.well-known/openid-configuration) and is able to fetch the JWKS from your IDP. It will then verify if the key used to sign the received token is in the JWKS and verify the signature with said key.

### PREST_JWT_JWKS
Instead of providing a well-known URL, you can set the JWKS yourself. It will have the same subsequent behavior as for the PREST_JWT_WELLKNOWN parameter.
Ex :  
```
{
  "keys": [
    {
      "n": "vcP1njqmR82oUndYDYxPF2HO-foOvDT8W8swcBYcgVQ0L1dPOlXz1lgQipJ0G7w9fXqilxsFHTs_MmXuGWQ2m3Wr16dpfPRFAN0ixj0ozzN_ZsF7OSdF6ZMZIJb3ObPt70i_emqEcd5ApITQ4PFvAhpZrBZCrW1z4aceoE1xC3rDV6qjA6pQr9u5A9J8Qj2a4LBG9mAPZCHQt4z50nhAVieKXhBoqKBoDplxj6yqzTTEUi1j-R0Z0NEJUIgt1_ErVsVMOWRKE6-BG39Zu9mvg2UAzxcitjOwa4n_spSgRleIliUJDNu-YpCEwj-5S_cX8DyHxCpvGp4ZBxPq1DUgqQ",
      "alg": "RS256",
      "use": "sig",
      "e": "AQAB",
      "kty": "RSA",
      "kid": "08bf5c3772dd4e7a727a101bf520f6575cac326f"
    }
  ]
}
```

### Unit tests fix
I changed the version of Go used in the docker image for the tests. It needed to be updated to at least 1.20 for the golang.org/x/exp module.

I also included tweaks to the docker-compose to be able to run it using podman instead of Docker.

## Summary by CodeRabbit

- **New Features**
	- Enhanced JWT handling by adding functionality to fetch JWKS from a well-known URL, improving app security.
- **Tests**
	- Expanded testing coverage for JWT configurations and key management, including RSA encryption handling.
- **Chores**
	- Updated Dockerfile to use `golang:1.22` and included `postgresql-client` installation for an updated development environment.

### CODE REVIEW

File: config/config_test.go [hunk 1]
```diff
@@ -94,6 +94,23 @@ func TestParse(t *testing.T) {
...
...8mmKg_GCsd5uKfJJ965zqSQu1CFn26YccTPp2doih4rykTGPVJdL5PVp3z4t9rTlahHbgCvv3E50yVK7LCNgtS9nmcZbD0meLqIZi3MoV0dBB_9C-qrEsevAIlPuXUmwtcbyDXOb1m7Xq_MPV_EASzoPYYjmk3k09zJ_p1EUTQ","e":"AQAB","x5c":["MIICmzCCAYMCBgGOLghSlzANBgkqhkiG9w0BAQsFADARMQ8wDQYDVQQDDAZtYXN0ZXIwHhcNMjQwMzExMTQ1OTQxWhcNMzQwMzExMTUwMTIxWjARMQ8wDQYDVQQDDAZtYXN0ZXIwggEiMA0GCSqGSIb3DQEBAQUAA4IBDwAwggEKAoIBAQDWrUjP57KGcJYIFEqCrTEqbpwXbMcCL9r4v3053CCdSCgZ+63lsFNTaEbH1HUgYqfSWgOMgK1TrgNUWk5pPuNjuggbgGxA3cS3kfo7deVWKbk1CAxhULqEoCt6qedXKVST5d9W+tA6i1WQZ8wV0BG4FovHW5sXEgdQ42LYyOCXadnyaYqD8YKx3m4p8kn3rnOpJC7UIWfbphxxM+nZ2iKHivKRMY9Ul0vk9WnfPi32tOVqEduAK+/cTnTJUrssI2C1L2eZxlsPSZ4uohmLcyhXR0EH/0L6qsSx68AiU+5dSbC1xvINc5vWbter8w9X8QBLOg9hiOaTeTT3Mn+nURRNAgMBAAEwDQYJKoZIhvcNAQELBQADggEBAIKBZNe4GmyfqRW6Ee8ai1umbstAmyK3W1kP2i0xxINTlvY2rwblV8UCrdyi3laD7zvZy1midZmpKqtZqWpiNigeZ5aUt76paYvdSl5TAuvZGDGoEAhmmECbnDSQKLp36rCn7NlrgiTDfZZ2PvIKZ3cXClzqXLF/iC6uGiKOgY5yOFOa5QgsfItpJmmxHtTzrRF70RVsbZCexB1Lt4bcId6Y3x2w7JNUjKIhf1RZ3QZx8+3xBM4cJ83h2J4nE0+IlFeAJL3VLGdeOk+z+FGMu2mYkxJwkxd9Wl2ubqrRcNy0t61Bgp3s40BgD10pzvawTXl7lEgabc/jzN2R0lcXmLo="],"x5t":"n5Y_Obidr330txi13j50zHzVbfg","x5t#S256":"f-Hrw_t_qUq86Ux0J2EckWVycuM3L_IjdOK6DW0DFoc"}]}`, cfg.JWTJWKS)
```
[the remark this answers is not in the document]
Comment (mcharest-mcn) [reply config/config_test.go:111]: This test is done in the middleware tests.

File: config/config_test.go [hunk 2]
```diff
@@ -94,6 +94,23 @@ func TestParse(t *testing.T) {
...
 	})
 
+	t.Run("PREST_JWT_WELLKNOWN", func(t *testing.T) {
+		//todo: Mock well-known config response
+		t.Setenv("PREST_JWT_WELLKNOWN", "https://accounts.google.com/.well-known/openid-configuration")
+		viperCfg()
+		cfg := &Prest{}
+		Parse(cfg)
+		require.Equal(t, "https://accounts.google.com/.well-known/openid-configuration", cfg.JWTWellKnown)
+	})
```
[the remark this answers is not in the document]
Comment (mcharest-mcn) [reply config/config_test.go:104]: I would have to mock a well-known URL response. Which I might need help with.

File: middlewares/config.go [hunk 3]
```diff
@@ -37,7 +37,7 @@ func initApp() {
 		if !config.PrestConf.Debug && config.PrestConf.EnableDefaultJWT {
 			MiddlewareStack = append(
 				MiddlewareStack,
-				JwtMiddleware(config.PrestConf.JWTKey, config.PrestConf.JWTAlgo))
+				JwtMiddleware(config.PrestConf.JWTKey, config.PrestConf.JWTJWKS))
```
Comment (arxdsilva) [inline middlewares/config.go:40]: could be better if this is configurable, so old users can still use that config and you can use this also

**Comment by mcharest-mcn** [comment]
I received a comment by @arxdsilva about middlewares/config.go where I should have kept
`JwtMiddleware(config.PrestConf.JWTKey, config.PrestConf.JWTAlgo)` but the modified function in middlewares.go manages the two situations. The JWTAlgo has disappeared because it was not used in the JwtMiddleware function anyway.

**Comment by mcharest-mcn** [comment]
Just a check for the integration of this pull request. Are there any more work to do?

File: config/config.go [hunk 4]
```diff
@@ -358,6 +371,50 @@ func parseDatabaseURL(cfg *Prest) {
...
+
+	var wellKnown map[string]interface{}
+	err = json.NewDecoder(r.Body).Decode(&wellKnown)
+	if err != nil {
+		log.Errorf("Failed to decode JSON: %v\n", err)
+		return
+	}
+
+	//Retrieve the JWKS from the endpoint
+	JWKSet, err := jwk.Fetch(context.Background(), wellKnown["jwks_uri"].(string))
```
Comment (arxdsilva) [inline config/config.go:401]: can you split this conversion into a variable? just to make sure that it never panics:

```go
uri, ok := wellKnown["jwks_uri"].(string)
...
JWKSet, err := jwk.Fetch(context.Background(),uri)
```

File: config/config.go [hunk 5]
```diff
@@ -271,7 +281,10 @@ func Parse(cfg *Prest) {
 	cfg.SingleDB = viper.GetBool("pg.single")
 	cfg.JWTKey = viper.GetString("jwt.key")
 	cfg.JWTAlgo = viper.GetString("jwt.algo")
+	cfg.JWTWellKnown = viper.GetString("jwt.wellknown")
```
Comment (arxdsilva) [inline config/config.go:284]: could you rename this to `JWTWellKnownURL` ?

File: config/config.go [hunk 6]
```diff
@@ -358,6 +371,50 @@ func parseDatabaseURL(cfg *Prest) {
...
+		log.Debugln("no JWT WellKnown url found, skipping")
+		return
+	}
+	if cfg.JWTJWKS != "" {
+		log.Debugln("JWKS already set, skipping")
+		return
+	}
+
+	// Call provider to obtain .well-known config
+	r, err := http.Get(cfg.JWTWellKnown)
```
Comment (arxdsilva) [inline config/config.go:386]: please add context to this call with a 5s timeout

File: middlewares/middlewares.go [hunk 7]
```diff
@@ -147,7 +148,41 @@ func JwtMiddleware(key string, algo string) negroni.Handler {
 			return
 		}
 		out := auth.Claims{}
-		if err := tok.Claims([]byte(key), &out); err != nil {
+		var rawkey interface{}
+
+		//Tester l'ensemble des types de clés...
```
Comment (arxdsilva) [inline middlewares/middlewares.go:153]: could you translate to english?

File: middlewares/middlewares.go [hunk 8]
```diff
@@ -147,7 +148,41 @@ func JwtMiddleware(key string, algo string) negroni.Handler {
...
+		if JWKSet == "" {
+			rawkey = []byte(key)
+		} else {
```
Comment (arxdsilva) [inline middlewares/middlewares.go:156]: could you invert this logic?
```go
// start setting this as default
rawkey = []byte(key)
if JWKSet != "" {
...
}

```

File: testdata/Dockerfile [hunk 9]
```diff
@@ -1,4 +1,4 @@
-FROM registry.hub.docker.com/library/golang:1.18
+FROM registry.hub.docker.com/library/golang:1.20
```
Comment (arxdsilva) [inline testdata/Dockerfile:1]: we can use go 1.22, also please update go.mod

**Comment by arxdsilva** [comment]
@mcharest-mcn thanks for the contribution, I feel that once these points are addressed we can merge it.

File: config/config.go (hunk 4 above)
Comment (mcharest-mcn) [reply config/config.go]: Done in latest commit.
https://github.com/prest/prest/pull/871/commits/fee30e7ae1162e49e1d88f7fa384c73b016fdc95

File: config/config.go (hunk 5 above)
Comment (mcharest-mcn) [reply config/config.go]: Done in latest commit.
https://github.com/prest/prest/pull/871/commits/fee30e7ae1162e49e1d88f7fa384c73b016fdc95

File: config/config.go (hunk 6 above)
Comment (mcharest-mcn) [reply config/config.go]: Done in latest commit.
https://github.com/prest/prest/pull/871/commits/fee30e7ae1162e49e1d88f7fa384c73b016fdc95

File: middlewares/middlewares.go (hunk 7 above)
Comment (mcharest-mcn) [reply middlewares/middlewares.go]: Removed in latest commit.
https://github.com/prest/prest/pull/871/commits/fee30e7ae1162e49e1d88f7fa384c73b016fdc95

File: middlewares/middlewares.go (hunk 8 above)
Comment (mcharest-mcn) [reply middlewares/middlewares.go]: Done in latest commit.
https://github.com/prest/prest/pull/871/commits/fee30e7ae1162e49e1d88f7fa384c73b016fdc95

File: testdata/Dockerfile (hunk 9 above)
Comment (mcharest-mcn) [reply testdata/Dockerfile]: Done in latest commit.
https://github.com/prest/prest/pull/871/commits/fee30e7ae1162e49e1d88f7fa384c73b016fdc95

**Review by arxdsilva — APPROVED** [review]
thanks @mcharest-mcn !

**Review by Noroz-77 — APPROVED** [review]

**Comment by Noroz-77** [comment]
c801f5508a38e766cd97dcd0df488dfed7f6b86b

**Pull request merged**